java-checks/src/main/java/org/sonar/java/checks/ConstantsShouldBeStaticFinalCheck.java

```diff
@@ -86,32 +85,36 @@ public void leaveNode(Tree tree) {
 
   private static boolean hasConstantInitializer(VariableTree variableTree) {
     ExpressionTree init = variableTree.initializer();
-    if (init != null) {
-      if (ExpressionUtils.skipParentheses(init).is(Tree.Kind.METHOD_REFERENCE)) {
-        MethodReferenceTree methodRef = (MethodReferenceTree) ExpressionUtils.skipParentheses(init);
-        if (isInstanceIdentifier(methodRef.expression())) {
-          return false;
-        }
-      }
-      if (init.is(Tree.Kind.NEW_ARRAY)) {
+    if (init == null) {
+      return false;
+    }
+
+    var deparenthesized = ExpressionUtils.skipParentheses(init);
+
+    if (deparenthesized.is(Tree.Kind.METHOD_REFERENCE)) {
+      MethodReferenceTree methodRef = (MethodReferenceTree) deparenthesized;
+      if (isInstanceIdentifier(methodRef.expression())) {
         return false;
       }
```

Comment:
This suggestion might be better readable instead. I think in new code, we generally should prefer the `instanceof` with pattern matching approach if we only check against a single `Kind` and need the instance. We should use the "old" way only if:

- We check against multiple kinds, i.e.:
  - `switch` instead of `if`, or
  - `is` with multiple arguments
- We check only against a single kind, but don't need the cast operation

```suggestion
    if (deparenthesized instanceof MethodReferenceTree methodRef && isInstanceIdentifier(methodRef.expression())) {
      return false;
    }
```